### PR TITLE
Fix cPickle deadlock when unpickling Job objects

### PR DESCRIPTION
**Pull request by joelarmstrong**
This is relevant to earlier issues like #1503. The root cause is that cPickle sometimes doesn't release the GIL when reading. When reading using the `read_file` function, cPickle correctly releases the GIL before performing a potentially-blocking operation:

https://github.com/python/cpython/blob/a61f5da54772b0ea6a7eccf21df08e61585ef712/Modules/cPickle.c#L552

But in the `readline_file` function, it doesn't:

https://github.com/python/cpython/blob/a61f5da54772b0ea6a7eccf21df08e61585ef712/Modules/cPickle.c#L591

This means that anytime the "file-like-object" that cPickle is reading from is a pipe filled from a different Python thread (i.e. in remote jobStores), we have to avoid using `readline_file` at all costs. Otherwise, the unpickling thread will stall waiting for data, but the thread that fills the other end of the pipe will never run, because the GIL is still held by the reading thread. We originally thought we worked around this by enforcing `cPickle.HIGHEST_PROTOCOL` on all pickle dumps. Pickle's protocol version 2 dumps most things in a binary format and not a line-based one, which means `readline_file` is almost never used (we thought it was never used). But unfortunately, `load_global` opcodes always cause `readline_file` to be used, even in the higher protocol versions:

https://github.com/python/cpython/blob/a61f5da54772b0ea6a7eccf21df08e61585ef712/Modules/cPickle.c#L4039

So sometimes, if you're unlucky, a load_global opcode will happen to be right on the end of the pipe-buffer (I believe that's 4KB on Linux), and there's a race between one thread reading past the end of the pipe, and the other writing more data into it. In my tests, this race gets lost roughly once in every 10K to 100K jobs for me, but it may depend heavily on the size of your Job objects and even the length of the names of the files/classes you use.

Basically, this PR solves the problem by never actually streaming the file when unpickling. Instead, it's fed into a special function which just reads all the data and calls `pickle.loads` on it, which can't stall because all the data is already in memory. (Why not avoid using readFileStream entirely and just use readFile?: Because this way we don't have to worry about cleaning up temporary files if there's an unpickling error.)

Unpickling the job objects is a special case, because it uses a custom unpickler, which doesn't have the `loads` method: in that case, we just read it into a temporary file and delete it after we're done or an exception is raised.

### CODE REVIEW

**Comment by DailyDreaming** [comment]
Jenkins, test this please.

**Comment by DailyDreaming** [comment]
Jenkins, test this please.

**Comment by DailyDreaming** [comment]
Thanks so much for getting this in @joelarmstrong !  It looks pretty clean.  The failing test is a bit odd (unable to find the s3 bucket?) and I haven't seen it before.  Not sure if it's a random fail or not.  Assuming tests pass, this looks ready to merge.

Also, thanks for the clear explanation of the problem with links.  Was hoping to browse python 3 and find that it properly released the GIL, but it seems pretty similar to the python 2.7 code: https://github.com/python/cpython/blob/master/Modules/_pickle.c#L1352 .  This looks like a great solution, thanks for submitting it!

**Comment by joelarmstrong** [comment]
Are these remaining failing tests known to be flaky? It looks like the static Azure test may have failed because the container took too long to start. (You guys might be getting throttled by Quay, I've run into similar issues). Not sure what's going on with the CWL conformance test.

Yeah, sorry for the verbosity. I just wanted to dump out all that information so other people know what to watch out for: I think no one on the current toil team was around when we first hit this bug. Yeah, sadly it was never fixed in any version of Python, as far as I can tell. (Honestly, I don't think anyone ever ran into it except us.)

**Comment by DailyDreaming** [comment]
The azure autoscaling test may be flaky.  Azure autoscaling was a little flaky for a time, but hasn't been failing in a while.  The cwl conformance tests have never been flaky since I've been here.

No, it was really valuable.  I didn't realize this had a history to it.  I'm glad you explained it and gave pointers to how the GIL release is controlled in the first function, and the second function lacking that control.

Not sure what's going on with the tests.  Will run again.  I suspect you're right and there may be a bottleneck somewhere causing some timing problem.

**Comment by DailyDreaming** [comment]
Jenkins, test this please.

**Comment by DailyDreaming** [comment]
The cwl conformance test oddly failed on #2179 as well, which should be a PR that has no effect at all on cwl performance.  I'm guessing they may have added new tests to the tarball that we fetch.

**Comment by DailyDreaming** [comment]
Jenkins, test this please.

**Comment by mr-c** [comment]
The CWL conformance tests are fetched via hash, so there is no way for use to sneak additional tests in 😁

**Comment by DailyDreaming** [comment]
@mr-c thanks!  I should have seen that.  It looks like the 3 failing cwl tests are all timing out (they all pass fine running it locally though).  A bunch of cloud tests also appear to be timing out too.  I suspect it might be #2157 that was added recently simply taking longer to delete things in general, in which case we should only apply the fix to nfs, rather than deletions in general.  I'll make a PR to test it.

**Comment by DailyDreaming** [comment]
Jenkins, test this please.

**Comment by DailyDreaming** [comment]
Flaky tests seem to be addressed and this is passing all but the known cwl test failure.  LGTM.
